src/pip/_internal/cache.py

```diff
@@ -157,10 +217,12 @@ def get_path_for_link(self, link):
 
         :param link: The link of the sdist for which this will cache wheels.
         """
-        parts = self._get_cache_path_parts(link)
-
-        # Store wheels within the root cache_dir
-        return os.path.join(self.cache_dir, "wheels", *parts)
+        path = self._get_path_for_link(link)
+        if not os.path.exists(path):
```

Comment:
I think this means that pip 20 will always keep using the previous cache structure.
We don't want to override `get_path_for_link` (used in reading and *writing* the cache) but only use `_get_path_for_link_legacy` in `_get_candidates`, when we are only reading the cache.

Reply:
@xavfernandez Right, sorry I skimmed too quickly on your previous comment.

> means that pip 20 will always keep using the previous cache structure

Only if `legacy_path` exists, i.e. when link was already cached in the previous cache structure. So yes with this approach new wheels with different tags built for the same link will go in the legacy structure if it existed before.
But new links will go in the new structure.

> in `_get_candidates`

Do you think its better to list both paths always, or only list the legacy path when the new one does not exist?

Reply:
> Only if legacy_path exists

You're right :+1: My bad, I read too quickly :sleepy: 
But ideally pip 20 should not be adding anything new in the legacy path anymore.

> Do you think its better to list both paths always, or only list the legacy path when the new one does not exist?

The way I imagined it was to only check the legacy path if the new one did not exist but it might even be better to check both if the user is using multiple interpreters.
Otherwise an install of a dependency with `pip>=20` on one interpreter one could deactivate an existing legacy cache of the same dependency for an other interpreter.

Reply:
@xavfernandez I've redone the last commit and added one for the removal news. I propose a 1 year transition period (i.e. until pip 21).